README.md
# SQL based data processing with declarative framework
This is a project developed with Python CDK for the solution SO0141 - SQL based ETL with a declarative framework.

## Prerequisites 
Python is needed in this project. Specifically, you will need version 3.6 or later. You can find information about downloading and installing Python [here](https://www.python.org/downloads/). Additionally you will need to have the Python package installer (pip) installed. See installation instructions [here](https://pypi.org/project/pip/).

If you use Windows, be sure Python is on your PATH. 


## Deploy Infrastructure

Clone the project.

```
$ git clone https://github.com/melodyyangaws/sql-based-etl.git
$ cd sql-based-etl

```

Build docker image and push to ECR via a bash script. 
arc-jupyter doesn't need image build, just pull it from public and push to ECR. The {{account_number}} in `deployment/environment.cfg` and sample job files under `source/app_resources/` will be updated by your input.

```
$ bash deployment/pull_and_push_ecr.sh <region> <account_number> <ecr_repo_name> <build_or_not>
 
# For example:
$ bash deployment/pull_and_push_ecr.sh 'us-west-2' 1234567 'arc' 1
$ bash deployment/pull_and_push_ecr.sh 'us-west-2' 1234567 'arc-jupyter'
```

This project is set up like a standard Python project. The `cdk.json` file tells where the application entry point is.The initialization process also creates a virtualenv within this project, stored under the .env
directory.  To create the virtualenv it assumes that there is a `python3`(or `python` for Windows) executable in your path with access to the `venv`
package. If for any reason the automatic creation of the virtualenv fails, you can create the virtualenv manually.

To manually create a virtualenv on MacOS and Linux:

``` 
$ python3 -m venv .env
```

After the init process completes and the virtualenv is created, you can use the following
step to activate your virtualenv.

```
$ source .env/bin/activate
```

Once the virtualenv is activated, you can install the required dependencies.

```
$ pip install -r source/requirements.txt
```

Install kubectl & jq tools

```
$ sudo curl --silent --location -o /usr/local/bin/kubectl \
   https://amazon-eks.s3.us-west-2.amazonaws.com/1.17.11/2020-09-18/bin/linux/amd64/kubectl
$ sudo chmod +x /usr/local/bin/kubectl

$ sudo yum -y install jq
OR
$ brew install jq
```

At this point you can now synthesize the CloudFormation template to be deployed.

```
$ cdk synth SparkOnEKS --require-approval never -c env=develop -o cdk.out

```

Finally deploy the stack. It takes two optional parameters `jhubuser` & `datalakebucket`.

```
# Scenario1: the recommended way is to deploy the default settings

$ cdk deploy SparkOnEKS --require-approval never -c env=develop

# Scenario2: give an arbitrary username as your Jupyter Hub login. 
# Otherwise, login with a default user.

$ cdk deploy SparkOnEKS --require-approval never -c env=develop --parameters jhubuser=<random_login_name>

# Scenario3: by default, the `datalakebucket` is set to the solution deployment S3 bucket, 
# if you want to use an existing bucket that contains real data, add the parameter to the command line
# an IAM role will be mapped to the existing S3 bucket for Jupyter Notebook and ETL jobs. 
# NOTE: the bucket must be in the same region as your infrastructure deployment.

$ cdk deploy SparkOnEKS --require-approval never -c env=develop --parameters jhubuser=<random_login_name> --parameters datalakebucket=<existing_datalake_bucket>

```
## Troubleshooting

1. If you see the issue `[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1123)`, most likely it means no default certificate authority for your Python installation on OSX. Refer to the [answer](https://stackoverflow.com/questions/52805115/certificate-verify-failed-unable-to-get-local-issuer-certificate) and installing `Install Certificates.command` should fix your local environment. Otherwise, use Cloud9 to deploy the CDK instead.

2. If an error says `SparkOnEKS failed: Error: This stack uses assets, so the toolkit stack must be deployed to the environment (Run "cdk bootstrap aws://YOUR_ACCOUNT_NUMBER/REGION")` , it means it is the first time you deploy an AWS CDK app into an environment (account/region), you’ll need to install a “bootstrap stack”. This stack includes resources that are needed for the toolkit’s operation. For example, the stack includes an S3 bucket that is used to store templates and assets during the deployment process.

You can use the cdk bootstrap command to install the bootstrap stack into an environment:

```
$ cdk bootstrap aws://YOUR_ACCOUNT_NUMBER/REGION -c develop
```


## Manually fix EKS node group security groups
Due to the issue https://github.com/aws/aws-cdk/issues/10884 , we will have to manually amend the SG for now.
1. Go to EC2 console and locate the instance `Spot-spark-on-eks-dev`, find an inbound rule which has the source from `eks-cluster-sg-spark-on-eks-dev` and port is 443, then change the `Type` to `All Traffic`. If the rule entry doesn't exist, add a new rule for it.
2. Find the managed node EC2 instance, which has the empty `Name` field and the value of the tag `eks:nodegroup-name` is `etl-job`. Similar to the previous step, change the inbound rule which has the source from `SparkOnEKS-EksClusterspotInstanceSecurityGroup` and port is 443, ie. modify the `Type` from `HTTPS` to `All Traffic`. 
3. restart any broken pods in Jupyter namespace, such as `hub` pod.

```
$ kubectl get pod -n jupyter
$ kubectl delete pod <hub-random-string>  -n jupyter
```

## Submit a Spark job on a web interface

After finished the deployment, we can start to `submit Spark job` on [Argo](https://argoproj.github.io/). Argo is an open source container-native workflow tool to orchestrate parallel jobs on Kubernetes. Argo Workflows is implemented as a Kubernetes CRD (Custom Resource Definition), which triggers time-based and event-based workflows specified by a configuration file.

Take a look at the [sample job](https://github.com/tripl-ai/arc-starter/tree/master/examples/kubernetes/nyctaxi.ipynb) developed in Jupyter Notebook.  It uses a Spark Wrapper called [ARC framework](https://arc.tripl.ai/) to create an ETL job in a codeless, declarative way. The opinionated standard approach enables rapid application deployment, simplifies data pipeline build. Additionally, it makes [self-service analytics](https://github.com/melodyyangaws/aws-service-catalog-reference-architectures/blob/customize_ecs/ecs/README.md) possible to the business.

In this example, we will extract the `New York City Taxi Data` from the [AWS Open Data Registry](https://registry.opendata.aws/nyc-tlc-trip-records-pds/), ie. a public s3 bucket `s3://nyc-tlc/trip data`, transform the data from CSV to parquet file format, followed by a SQL-based data validation step, to ensure the typing transformation is done correctly. Finally, query the optimized data filtered by a flag column.

1. Find your Argo dashboard URL from the deployment output, something like this:
![](/images/0-argo-uri.png)

2. Go to the ARGO dashboard, click on the `SUBMIT NEW WORKFLOW` button.
![](/images/1-argoui.png)

3. Replace the existing manifest by the following job definition, click `SUBMIT`.

```
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: nyctaxi-job-
  namespace: spark
spec:
  serviceAccountName: arcjob
  entrypoint: nyctaxi
  templates:
  - name: nyctaxi
    dag:
      tasks:
        - name: step1-query
          templateRef:
            name: spark-template
            template: smallJob
            clusterScope: true   
          arguments:
            parameters:
            - name: jobId
              value: nyctaxi 
            - name: image
              value: ghcr.io/tripl-ai/arc:latest
            - name: configUri
              value: https://raw.githubusercontent.com/tripl-ai/arc-starter/master/examples/kubernetes/nyctaxi.ipynb
            - name: parameters
              value: "--ETL_CONF_DATA_URL=s3a://nyc-tlc/trip*data --ETL_CONF_JOB_URL=https://raw.githubusercontent.com/tripl-ai/arc-starter/master/examples/kubernetes"

```
![](/images/2-argo-submit.png)

4. Click a pod (dot) on the dashboard to examine the job status and application logs.
![](/images/3-argo-log.png)


## Submit a Spark job via command line

We have included a second job manifest file and a Jupyter notebook, as an example of a complex Spark job to solve a real-world data problem. Let's submit it via a command line this time. 
<details>
<summary>manifest file</summary>
The [manifest file](/source/app_resources/scd2-job.yaml) defines where the Jupyter notebook file (job configuration) and input data are. 
</details>
<details>
<summary>Jupyter notebook</summary>
The [Jupyter notebook](/deployment/app_code/job/scd2_job.ipynb) specifies what exactly need to do in a data pipeline.
</details>

In general, a parquet file is immutable in a Data Lake. This example will demonstrate how to address the challenge and process data incrementally. It uses `Delta Lake`, which is an open source storage layer on top of parquet file, to bring the ACID transactions to Apache Spark and modern big data workloads. In this solution, we will build up a table to meet the [Slowly Changing Dimension Type 2](https://www.datawarehouse4u.info/SCD-Slowly-Changing-Dimensions.html) requirement, and prove that how easy it is when ETL with a SQL first approach implemented in a configuration-driven architecture.


1.Open the `scd2 workflow job` manifest file, replace the S3 bucket by a correct name at each task sections. Either the code bucket name from your deployment output, or an existing datalake bucket passed in as a deployment parameter.

```
$ vi source/app_resources/scd2-workflow-job.yaml
```
![](/images/4-cfn-output.png)

![](/images/3-scd-bucket.png)

2.Before running command lines, you need to connect to the Amazon EKS cluster firstly. Get the connection command from the deployment output, something like this:

```
aws eks update-kubeconfig --name <EKS_cluster_name> --region <region> --role-arn arn:aws:iam::<account_number>:role/<role_name>
```
![](/images/0-eks-config.png)

3.Submit Arc Spark job

```
$ kubectl apply -f source/app_resources/scd2-workflow-job.yaml

# Delete before submit the same job again
$ kubectl delete -f source/app_resources/scd2-workflow-job.yaml
$ kubectl apply -f source/app_resources/scd2-workflow-job.yaml
```

<details>
<summary> 
Alternatively, submit the job without the extra deletion step by [Argo CLI](https://www.eksworkshop.com/advanced/410_batch/install/). 
</summary> 

```
$ argo submit source/app_resources/scd2-workflow-job.yaml -n spark --watch

# terminate the job that is running
$ argo delete scd2-job-<random_string> -n spark

```
*** Make sure to comment out a line in the manifest file before your submission. ***
![](/images/2-comment-out.png)
![](/images/2-argo-scdjob.png)
</details>

4.Go to your Argo dashboard by running the following command. Or copy the URL directly from your deployment output.

```
$ ARGO_URL=$(kubectl -n argo get ingress argo-server --template "{{ range (index .status.loadBalancer.ingress 0) }}{{ . }}{{ end }}")
$ echo ARGO DASHBOARD: http://${ARGO_URL}:2746
```

5.Check the job status and applications logs on the dashboard.
![](/images/3-argo-log.png)


## Create / test Spark job in Jupyter notebook

Apart from orchestrating Spark jobs with a declarative approach, we introduce a codeless, configuration-driven design for increasing data process productivity, by leveraging an open-source [data framework ARC](https://arc.tripl.ai/) for a SQL-centric ETL solution. We take considerations of the needs and expected skills from our customers in data, and accelerate their interaction with ETL practice in order to foster simplicity, while maximizing efficiency.

1.Login to Jupyter Hub. The default username is `sparkoneks`, or use your own login passed in as a deployment parameter. The default password is `supersecret!`

```
$JHUB_URL=$(kubectl -n jupyter get ingress -o json | jq -r '.items[] | [.status.loadBalancer.ingress[0].hostname] | @tsv')
$echo jupyter hub login: http://${JHUB_URL}:8000

```

2.Start the default development environment. Use a big Jupyter Hub instance to author your ETL job if you prefer.

![](/images/3-jhub-login.png)

3.Upload a sample Arc-jupyter notebook from `deployment/app-code/job/scd2_job.ipynb`
![](/images/3-jhub-upload.png =80x100)

![](/images/3-jhub-open-notebook.png)

4.Execute each block and observe the result.

NOTE: the variable `${ETL_CONF_DATALAKE_LOC}` is set to a code bucket created by the deployment, and the IAM role attached to the Jupyter Hub is restricted to the S3 bucket. If you would like to connect to an existing bucket that contains real data, re-run the deployment with your `datalakebucket` name as a parameter, or simply add the bucket ARN to the IAM role 'SparkOnEKS-EksClusterjhubServiceAcctRole'.

```
cdk deploy SparkOnEKS -c env=develop --require-approval never -c env=develop --parameters datalakebucket=<real_data_bucket>
```

## Submit a native Spark job with Spot instance

As an addition of the solution, to meet customer's preference of running native Spark jobs, we will demonstrate how easy to submit a job in EKS. In this case, the Jupyter notebook still can be used, as an interactive development environment for PySpark apps. 

In Spark, driver is a single point of failure in data processing. If driver dies, all other linked components will be discarded as well. To achieve the optimal performance and cost, we will run the driver on a reliable managed EC2 instance on EKS, and the rest of executors will be on spot instances.

1.[OPTIONAL] Run a dummy container to check source code files.

```
$ kubectl run --generator=run-pod/v1 jump-pod --rm -i --tty --serviceaccount=nativejob --namespace=spark --image <account_number>.dkr.ecr.<region>.amazonaws.com/arc:latest sh

# after login to the container, validate the files
# make sure the service account is `nativejob` and the lifecycle value is `Ec2Spot` in the executor template
sh-5.0# ls 
sh-5.0# cat executor-pod-template.yaml
sh-5.0# exit

```
2.Copy the S3 bucket name from the deployment output

![](/images/4-cfn-output.png)

3.Modify the job manifest file `native-spark-job.yaml` 
by changing the destination bucket.

![](/images/4-spark-output-s3.png)


4.Submit the native Spark job

```
$ kubectl apply -f source/app_resources/native-spark-job.yaml

```
5.Go to SparkUI to check the job progress and performance

```
$ driver=$(kubectl get pod -n spark -l spark-role=driver -o json | jq -r '.items[] | [.metadata.name] | @tsv')
$ kubectl port-forward $driver 4040:4040 -n spark

# go to your web browser, type in `localhost:4040`

```
6. Examine the auto-scaling and multiAZs

```
# watch the number of EC2 instances. The job requests 5 executors with 5 new Spot instances. 
# the auto-scaling will be triggered across multiple AZs.

$ kubectl get node --label-columns=lifecycle,topology.kubernetes.io/zone
$ kubectl get pod -n spark
```
![](/images/4-auto-scaling.png)

## Useful commands

 * `cdk ls`          list all stacks in the app
 * `cdk synth`       emits the synthesized CloudFormation template
 * `cdk deploy`      deploy this stack to your default AWS account/region
 * `cdk diff`        compare deployed stack with current state
 * `cdk docs`        open CDK documentation
 * `cdk destroy`     delete the stack deployed earlier
 * `kubectl get pod -n spark`                         list all jobs running in the Spark namespace
 * `kubectl get svc -n argo`                          get argo dashboard URL
 * `kubectl get svc -n jupyter`                       get Jupyterhub's URL
 * `argo submit source/app_resources/spark-job.yaml`  submit a spark job from a manifest file
 * `argo list --all-namespaces`                       show jobs from all namespaces
 * `kubectl delete pod --all -n spark`                delete all jobs submitted in the Spark namespace
* `kubectl apply -f source/app_resources/spark-template.yaml` submit a reusable job template for Spark applications



CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.0.0] - 2020-10-05
### Added
- All files, initial version


source/base_infra_stack.py
from aws_cdk import (
    core, 
    aws_eks as eks,
    aws_ec2 as ec2,
    aws_efs as efs,
    aws_iam as iam,
    aws_s3 as s3,
    aws_s3_deployment as s3deploy
)

from lib.network_sg import NetworkSgConst
from lib.iam_roles import IamConst
from lib.manifest_reader import *


class BaseEksInfraStack(core.Stack):

    # @property
    # def appcode_bucket(self):
    #     return self._artifact_bucket.bucket_name 

    # @property    
    # def eks_cluster(self):
    #     return self._my_cluster

    def __init__(self, scope: core.Construct, id: str, eksname: str, admin_name: str, **kwargs) -> None:
        super().__init__(scope, id, **kwargs)

        eks_vpc = NetworkSgConst(self,'network-sg', eksname).vpc
        iam_role = IamConst(self,'iam_roles', eksname, admin_name)

# //**********************************************************************//
# //******************* Upload app code to S3 bucket *********************//
# //*********************************************************************//
        self._artifact_bucket = s3.Bucket(self, "codeBucket",encryption=s3.BucketEncryption.KMS_MANAGED)

        s3deploy.BucketDeployment(self, "DeployCode",
            sources=[s3deploy.Source.asset("deployment/app_code")],
            destination_bucket=self._artifact_bucket,
        )

        code_bucket = self._artifact_bucket.bucket_name

# //**********************************************************************//
# //******************* EKS CLUSTER WITH NO NODE GROUP *******************//
# //*********************************************************************//

        self._my_cluster = eks.Cluster(self,'EksCluster',
                vpc= eks_vpc,
                # role= _iam_role.control_plane_role
                # security_group= NetworkSgConst.control_plane_sg,
                cluster_name=eksname,
                output_cluster_name=True,
                masters_role= iam_role.admin_role,
                version= eks.KubernetesVersion.V1_18,
                endpoint_access= eks.EndpointAccess.PUBLIC_AND_PRIVATE,
                default_capacity=0
        )

# //***********************************************************************//
# //********************* MANAGED NODE GROUP IN EKS *************************//
# //***************  compute resource to run Spark jobs *******************//
# //*********************************************************************//

        _managed_node = self._my_cluster.add_nodegroup_capacity('managed-nodegroup',
            nodegroup_name = 'etl-job',
            node_role = iam_role.managed_node_role,
            desired_size = 1,
            max_size = 5,
            min_size = 1,
            disk_size = 50,
            instance_type = ec2.InstanceType('r5.xlarge'),
            labels = {'app':'spark', 'lifecycle':'OnDemand'},
            subnets = ec2.SubnetSelection(subnet_type=ec2.SubnetType.PRIVATE,one_per_az=True)
        )
        core.Tags.of(_managed_node).add('Name', 'Managed-'+eksname)
# //****************************************************************************************//
# //*********************** Add Spot NodeGroup to EKS *************************************//
# //************* Run Spark driver on reliable on-demand, exectutor on spot **************//
# //************ CDK automatically installs the spot interrupt handler daemon ***********//
# //******************  to handle EC2 Spot Instance Termination Notices.****************//
# //***********************************************************************************//

        _spot_node = self._my_cluster.add_auto_scaling_group_capacity('spot',
            instance_type=ec2.InstanceType('r4.xlarge'),
            bootstrap_options={
                "kubelet_extra_args": "--node-labels app=spark",
            },
            min_capacity=1,
            max_capacity=10,
            spot_price='1'
        )
        # add to cluster auto scaler
        core.Tags.of(_spot_node).add('Name', 'Spot-'+eksname)
        core.Tags.of(_spot_node).add('k8s.io/cluster-autoscaler/enabled', 'true')
        core.Tags.of(_spot_node).add('k8s.io/cluster-autoscaler/'+eksname, 'owned')
# //**************************************************************************************//
# //********************* Add Farget Node to EKS ****************************************//
# // *************** Jupyter acts as an IDE for develop/test ETL job *******************//
# //******* Easy to maintain when it comes to the multi-tenancy, personaliation *******//
# //**********************************************************************************//

        # _fargate_node=self._my_cluster.add_fargate_profile('FargateProfile',
        #     selectors=[{"namespace":"jupyter"}],
        #     fargate_profile_name='jupyterhub',
        #     pod_execution_role=iam_role.fargate_role,
        #     subnet_selection=ec2.SubnetSelection(subnet_type=ec2.SubnetType.PRIVATE, one_per_az=True)
        # )
        # core.Tags.of(_fargate_node).add('k8s.io/cluster-autoscaler/enabled', 'true')
        # core.Tags.of(_fargate_node).add('k8s.io/cluster-autoscaler/'+eksname, 'owned')
# //************************************v*************************************************************//
# //***************************** SERVICE ACCOUNT, RBAC and IAM ROLES *******************************//
# //**** Associating AWS IAM role to K8s Service Account to provide fine-grain security control ****//
# //***********************************************************************************************//

       # Setup service account in a new namespace Spark
        _etl_ns = self._my_cluster.add_manifest('SparkNamespace',{
                "apiVersion": "v1",
                "kind": "Namespace",
                "metadata": { 
                    "name": "spark",
                    "labels": {"name":"spark"}
                }
            }
        )    
        _etl_sa = self._my_cluster.add_service_account('ETLSa', name='arcjob', namespace='spark')
        _etl_sa.node.add_dependency(_etl_ns)

        _setting = {"{{MY_SA}}": _etl_sa.service_account_name}
        _etl_k8s_rb = eks.KubernetesManifest(self,'ETLRoleBinding',
            cluster=self._my_cluster,
            manifest=loadYamlReplaceVarLocal('../app_resources/etl-rbac.yaml', fields= _setting, multi_resource=True)
        )
        _etl_k8s_rb.node.add_dependency(_etl_sa)

        # Associate AWS IAM role to K8s Service Account
        datalake_bucket = core.CfnParameter(self, "datalakebucket", type="String",
            description="Your existing S3 bucket name to be accessed by Jupyter Notebook and ETL job",
            default=""
        )
        _etl_bucket = code_bucket if datalake_bucket.value_as_string=="" else datalake_bucket.value_as_string
        _etl_s3_statements = loadYamlReplaceVarLocal('../app_resources/etl-iam-role.yaml', 
            fields={
                "{{codeBucket}}": _etl_bucket
                }
        )
        for statmnt in _etl_s3_statements:
            _etl_sa.add_to_principal_policy(iam.PolicyStatement.from_json(statmnt))

        # Cluster Auto-scaler service account
        _scaler_sa = self._my_cluster.add_service_account('AutoScalerSa', 
            name='cluster-autoscaler', 
            namespace='kube-system'
        )  
        _scaler_role = loadYamlLocal('../app_resources/autoscaler-iam-role.yaml')
        for statmt in _scaler_role:
            _scaler_sa.add_to_principal_policy(iam.PolicyStatement.from_json(statmt))

        # ALB Ingress service account
        _alb_sa = self._my_cluster.add_service_account('ALBServiceAcct', 
            name='alb-ingress-controller',
            namespace='kube-system'
        )
        _alb_role = loadYamlLocal('../app_resources/alb-iam-role.yaml')
        for statmt in _alb_role:
            _alb_sa.add_to_principal_policy(iam.PolicyStatement.from_json(statmt))

# //*************************************************************************************//
# //********************* Enable Cluster Autoscaler ************************************//
# //********************* add nodes to node group when it is needed *******************//
# //**********************************************************************************//
 
        _var_mapping = {
            "{{region_name}}": self.region, 
            "{{cluster_name}}": eksname, 
        }
        _scaler_chart = self._my_cluster.add_helm_chart('ClusterAutoScaler',
            chart='cluster-autoscaler-chart',
            repository='https://kubernetes.github.io/autoscaler',
            release='nodescaler',
            create_namespace=False,
            namespace='kube-system',
            values=loadYamlReplaceVarLocal('../app_resources/autoscaler-values.yaml',_var_mapping)
        )
        _scaler_chart.node.add_dependency(_scaler_sa) 

# //*************************************************************************************//
# //************************ CONTAINER INSIGHT (CLOUDWATCH LOG) ************************//
# //**********************************************************************************//

        _cw_log = eks.KubernetesManifest(self,'ContainerInsight',
            cluster=self._my_cluster, 
            manifest=loadYamlReplaceVarRemotely('https://raw.githubusercontent.com/aws-samples/amazon-cloudwatch-container-insights/latest/k8s-deployment-manifest-templates/deployment-mode/daemonset/container-insights-monitoring/quickstart/cwagent-fluentd-quickstart.yaml', 
                    fields=_var_mapping,
                    multi_resource=True
            )
        )

# //*********************************************************************//
# //*********************** ALB INGRESS CONTROLLER **********************//
# //*********************************************************************//

        # Deploy the ALB Ingress Controller from Helm chart
        dataMap = {
            "{{region_name}}": self.region, 
            "{{cluster_name}}": eksname, 
            "{{vpc_id}}": eks_vpc.vpc_id
        }
        _alb_chart = self._my_cluster.add_helm_chart('ALBChart',
            chart='aws-alb-ingress-controller',
            repository='http://storage.googleapis.com/kubernetes-charts-incubator',
            release='alb',
            create_namespace=False,
            namespace='kube-system',
            values=loadYamlReplaceVarLocal('../app_resources/alb-values.yaml',dataMap)
        )
        _alb_chart.node.add_dependency(_alb_sa)

# # # Add EFS for S3A staging commiter for faster S3 access
#         _csi_driver = self._my_cluster.add_manifest('efs-driver', 
#             loadYamlRemotely('https://raw.githubusercontent.com/kubernetes-sigs/aws-efs-csi-driver/deploy/kubernetes/overlays/stable/ecr/?ref=release-1.0')
#         )
#         _k8s_efs = efs.FileSystem(self,'create-efs',
#             vpc=eks_vpc,encrypted=True,
#             lifecycle_policy=efs.LifecyclePolicy.AFTER_14_DAYS,
#             performance_mode=efs.PerformanceMode.MAX_IO,
#             throughput_mode=efs.ThroughputMode.BURSTING,
#         )
#         _efs_mount_target = efs.CfnMountTarget(self,'mountEFS',
#             file_system_id=_k8s_efs.file_system_id,
#             security_groups=NetworkSgConst.efs_sg,
#             subnet_id=ec2.SubnetSelection(subnet_type=ec2.SubnetType.PUBLIC, one_per_az=True)

#         )
#         self._my_cluster.add_manifest('efs-storage',
#             manifest=loadYamlLocal('../app_resources/efs-storageclass.yaml')
#        )
#         _pv= self._my_cluster.add_manifest('pv-workdir-s3a',
#         manifest={
#             "apiVersion": "v1"
#             "kind": "PersistentVolume"
#             "metadata":{"name": "workdir-pv"}
#             "spec":{"capacity":{"storage": "5Gi"}}
#             "volumeMode":"Filesystem"
#             "accessModes":[{
#                 "ReadWriteMany"
#             }]
#             "persistentVolumeReclaimPolicy": "Retain"
#             "storageClassName":  "efs-sc"
#             "csi":{
#                 "driver": "efs.csi.aws.com",         
#                 "volumeHandle": _k8s_efs.file_system_id
#             }
#         }
#        _pv.node.add_dependency(_k8s_efs)

# //*************************************************************************************//
# //************************** Setup ARGO WORKFLOW *************************************//
# //******* K8s Native WF tool orchestrate containerized batch & streaming jobs ********//
# //***********************************************************************************//
        
        # Installation
        _argo_install = self._my_cluster.add_helm_chart('ARGOChart',
            chart='argo',
            repository='https://argoproj.github.io/argo-helm',
            release='argo',
            namespace='argo',
            create_namespace=True,
            values=loadYamlLocal('../app_resources/argo-values.yaml')
        )

        # Create a Spark workflow template, manually submit jobs based on the template later.
        _submit_tmpl = self._my_cluster.add_manifest('SubmitSparkWrktmpl',
            loadYamlLocal('../app_resources/spark-template.yaml')
        )
        _submit_tmpl.node.add_dependency(_argo_install)

# //*********************************************************************//
# //***************************** Setup Jupyter **************************//
# //*********************************************************************//

        _jhub_install= self._my_cluster.add_helm_chart('JHubChart',
            chart='jupyterhub',
            repository='https://jupyterhub.github.io/helm-chart',
            release='jhub',
            version='0.10.2',
            namespace='jupyter',
            create_namespace=True,
            values=loadYamlReplaceVarLocal('../app_resources/jupyter-config.yaml', 
                fields={
                    "{{codeBucket}}": _etl_bucket
                    }
                )
        )
        _expose_hub = self._my_cluster.add_manifest('JHubIngress',
            loadYamlLocal('../app_resources/jupyter-ingress.yaml')
        )
        _expose_hub.node.add_dependency(_jhub_install)

        # Create a login user name determined by a CFN input param
        login_name = core.CfnParameter(self, "jhubuser", type="String",
            description="The username login to jupyter hub",
            default="sparkoneks"
        )
        _hub_sa = self._my_cluster.add_service_account('jhubServiceAcct', 
            name=login_name.value_as_string,
            namespace='jupyter'
        )
        _hub_sa.node.add_dependency(_jhub_install)

        _setting = {"{{MY_SA}}": _hub_sa.service_account_name}
        _jhub_user_rb = self._my_cluster.add_manifest('hubRoleBinding',loadYamlReplaceVarLocal('../app_resources/jupyter-rbac.yaml',fields= _setting))
        _jhub_user_rb.node.add_dependency(_hub_sa)

        for statmnt in _etl_s3_statements:
            _hub_sa.add_to_principal_policy(iam.PolicyStatement.from_json(statmnt))

# # //*************************************************************************************//
# # //************ Create service account for native Spark jobs  **************************//
# # //***********************************************************************************//
        _spark_sa = self._my_cluster.add_service_account('nativeSparkSa',
            name='nativejob',
            namespace='spark'
        )
        _spark_sa.node.add_dependency(_etl_ns)

        _setting = {"{{MY_SA}}": _spark_sa.service_account_name}
        _spark_rb = self._my_cluster.add_manifest('sparkRoleBinding',
            loadYamlReplaceVarLocal('../app_resources/native-spark-rbac.yaml',fields= _setting)
        )
        _spark_rb.node.add_dependency(_spark_sa)

        _setting={"{{codeBucket}}": code_bucket}
        _spark_iam = loadYamlReplaceVarLocal('../app_resources/native-spark-iam-role.yaml',fields=_setting)
        for statmnt in _spark_iam:
            _spark_sa.add_to_principal_policy(iam.PolicyStatement.from_json(statmnt))

# # # # //*********************************************************************//
# # # # //**************** Cloud9 for eksctl in private subnet ****************//
# # # # //*********************************************************************//
# # # # 
# # #         # # create a cloud9 ec2 environment in a new VPC
# # #         # c9env = cloud9.Ec2Environment(self, 'Cloud9Env',
# # #         #     vpc=NetworkSgConst.vpc,
# # #         #     instance_type=ec2.InstanceType('t3.small'),
# # #         #     subnet_selection=ec2.SubnetSelection(subnet_type=ec2.SubnetType.PUBLIC)
# # #         # )

# # #         # # print the Cloud9 IDE URL in the output
# # #         # core.CfnOutput(self, 'URL', value=c9env.ide_url)


# # # //*********************************************************************//
# # # //*************************** Deployment Output ***********************//
# # # //*********************************************************************//
        argo_url=eks.KubernetesObjectValue(self, 'argoALB',
            cluster=self._my_cluster,
            json_path='.status.loadBalancer.ingress[0].hostname',
            object_type='ingress',
            object_name='argo-server',
            object_namespace='argo',
            timeout=core.Duration.minutes(10)
        )
        argo_url.node.add_dependency(_argo_install)
        core.CfnOutput(self,'ARGO_URL', value='http://'+ argo_url.value + ':2746')
        
        jhub_url=eks.KubernetesObjectValue(self, 'jhubALB',
            cluster=self._my_cluster,
            json_path='.status.loadBalancer.ingress[0].hostname',
            object_type='ingress',
            object_name='jupyterhub',
            object_namespace='jupyter',
            timeout=core.Duration.minutes(10)
        )
        jhub_url.node.add_dependency(_expose_hub)
        core.CfnOutput(self,'JUPYTER_URL', value='http://'+ jhub_url.value + ':8000')

        core.CfnOutput(self,'CODE_BUCKET', value=code_bucket)